### DISCUSSION TITLE
PID Setpoint Tracking

### DISCUSSION THREAD

**Discussion by isabido** [arendst/Tasmota#20382] [Ideas]
HI Theo!

I am designing a system at home where I need all photovoltaic production to be used to heat water.

What I need is, if the solar panels are generating 1kw (this measurement is extracted by mqtt from the inverter) I need to divert all that power to a resistor controlled by a dimmer with a 0-3.3v signal input.

Regulator 25A
https://es.aliexpress.com/item/20000003997748.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp

0-3.3v to 4-20mA converter
https://es.aliexpress.com/item/1005002960048002.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp

In this case the set point is variable depending on the energy produced by the solar panels.

I think this is called PID Setpoint Tracking where the value SP=PV.

Here I leave the technical explanation.
https://control.com/textbook/closed-loop-control/practical-pid-controller-features/

Would it be possible to perform this function with Tasmota?

Thanks in advance.

**Comment by barbudor** [comment]
Hello, Tasmota has a PID function which is barely described here : https://tasmota.github.io/docs/PID-Control/
May be you should look at the code as recommended

I have no idea how it works but I feel like, as the Thermostat feature, it is intended to drive a relay, not output a voltage.
Beside, Tasmota control over DAC is limited and mostly available only through [Berry DAC gpios](https://tasmota.github.io/docs/Berry/#dac-gpios)

So at the end you will porbably have to implement your PID algorithm in Berry
